### PR TITLE
Fix issue 12471

### PR DESCRIPTION
**Pull request by ankenyr**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This is to fix https://github.com/ytdl-org/youtube-dl/issues/12471
Currently with the use of [tempfile](https://github.com/ytdl-org/youtube-dl/blob/master/youtube_dl/utils.py#L1826) on Linux it will create a file with only the user read write bit set for user and all other bits off. When renaming the file it keeps this restrictive permission setting. From my testing this seems to not have an affect in Windows.

### CODE REVIEW

**Comment by ankenyr** [comment]
Seems I had some leftovers from years ago I forgot about. Removing those

File: youtube_dl/utils.py [hunk 1]
```diff
@@ -1835,6 +1836,12 @@ def write_json_file(obj, fn):
                 os.unlink(fn)
             except OSError:
                 pass
+        os.chmod(tf.name,
+                 stat.S_IRUSR |
+                 stat.S_IWUSR |
+                 stat.S_IRGRP |
+                 stat.S_IROTH)
+
```
Comment (dstftw) [inline youtube_dl/utils.py:1844]: 1. No umask respected.
2. There is no `os.chmod` in python 3.2 according to python [docs](https://docs.python.org/3/library/os.html#os.chmod).
3. flake8.

Comment (ankenyr) [reply youtube_dl/utils.py]: umask should be good now.
flake8 no longer complaines (sorry about that one!)
Not sure what you mean about os.chmod not being in python3.2?
The link takes me to 3.8 and the python 3.2 docs [clearly show its there](https://docs.python.org/3.2/library/os.html)
If I remember when I last checked tf.name returned a string object so that should work.

**Comment by ankenyr** [comment]
I believe this is good to be looked at again. @dstftw 
Please let me know if I am missing something.

**Comment by ankenyr** [comment]
@dstftw I am hoping this can be reviewed and included in the next release.

File: youtube_dl/utils.py [hunk 2]
```diff
@@ -1835,6 +1836,17 @@ def write_json_file(obj, fn):
...
+        mask = os.umask(0)
+        os.umask(mask)
+        os.chmod(tf.name,
+                 (stat.S_IRUSR
+                  | stat.S_IWUSR
+                  | stat.S_IRGRP
+                  | stat.S_IWGRP
+                  | stat.S_IROTH
+                  | stat.S_IWOTH)
+                 & ~mask)
```
Comment (dstftw) [inline youtube_dl/utils.py:1848]: This must be in a separate try-except so that it does not break renaming to correct name if any of these statements fails.

File: youtube_dl/utils.py [hunk 3]
```diff
@@ -1835,6 +1836,17 @@ def write_json_file(obj, fn):
...
                 pass
+        mask = os.umask(0)
+        os.umask(mask)
+        os.chmod(tf.name,
+                 (stat.S_IRUSR
+                  | stat.S_IWUSR
+                  | stat.S_IRGRP
+                  | stat.S_IWGRP
+                  | stat.S_IROTH
+                  | stat.S_IWOTH)
```
Comment (dstftw) [inline youtube_dl/utils.py:1847]: Just `0o666`.

File: youtube_dl/utils.py [hunk 4]
```diff
@@ -1835,6 +1836,17 @@ def write_json_file(obj, fn):
...
+                 & ~mask)
+
```
Comment (dstftw) [inline youtube_dl/utils.py:1849]: Remove unnecessary whitespace.

Comment (ankenyr) [reply youtube_dl/utils.py:1849]: done

File: youtube_dl/utils.py [hunk 5]
```diff
@@ -1835,6 +1835,12 @@ def write_json_file(obj, fn):
                 os.unlink(fn)
             except OSError:
                 pass
+        try:
+            mask = os.umask(0)
+            os.umask(mask)
+            os.chmod(tf.name, 0o666 & ~mask)
+        except OSError:
```
Comment (dstftw) [inline youtube_dl/utils.py:1842]: Are you sure no other exceptions possible here?

Comment (ankenyr) [reply youtube_dl/utils.py:1842]: I believe that this module will only raise OSError per the docs
https://docs.python.org/3/library/os.html

> Note All functions in this module raise OSError (or subclasses thereof) in the case of invalid or inaccessible file names and paths, or other arguments that have the correct type, but are not accepted by the operating system.
